src/cmd/ksh93/tests/readonly.sh

```diff
@@ -0,0 +1,338 @@
+		res='typeset -p fb'
+		exp='FB_t fb=(x=foo;y=bam;typeset -r z=bar)'
+	)
+	(
+		ini='typeset -ra arr=((a b c) 1)'
+		chg='arr+=(2)'
+		res='typeset -p arr'
+		exp='typeset -r -a arr=((a b c) 1)'
+	)
+	(
+		ini='typeset -ra arr=((a b c) 1)'
+		chg='arr[0]+=(d)'
+		res='typeset -p arr'
+		exp='typeset -r -a arr=((a b c) 1)'
+	)
+	(
+		ini='typeset -ra -i arr=((10 11 12) 3 4)'
+		chg='(( arr[1] += 2 ))'
+		res='typeset -p arr'
+		exp='typeset -r -a -i arr=((10 11 12) 3 4)'
+	)
+	(
+		ini='typeset -ra -i arr=((10 11 12) 3 4)'
+		chg='(( arr[1] -= 2 ))'
+		res='typeset -p arr'
+		exp='typeset -r -a -i arr=((10 11 12) 3 4)'
+	)
+	(
+		ini='typeset -ra -i arr=((10 11 12) 3 4)'
+		chg='(( arr[1] *= 2 ))'
+		res='typeset -p arr'
+		exp='typeset -r -a -i arr=((10 11 12) 3 4)'
+	)
+	(
+		ini='typeset -ra -i arr=((10 11 12) 3 4)'
+		chg='(( arr[1] /= 2 ))'
+		res='typeset -p arr'
+		exp='typeset -r -a -i arr=((10 11 12) 3 4)'
+	)
+	(
+		ini='typeset -ra -i arr=((10 11 12) 3 4)'
+		chg='(( arr[0][1] += 2 ))'
+		res='typeset -p arr'
+		exp='typeset -r -a -i arr=((10 11 12) 3 4)'
+	)
+	(
+		ini='typeset -ra -i arr=((10 11 12) 3 4)'
+		chg='(( arr[0][1] -= 2 ))'
+		res='typeset -p arr'
+		exp='typeset -r -a -i arr=((10 11 12) 3 4)'
+	)
+	(
+		ini='typeset -ra -i arr=((10 11 12) 3 4)'
+		chg='(( arr[0][1] *= 2 ))'
+		res='typeset -p arr'
+		exp='typeset -r -a -i arr=((10 11 12) 3 4)'
+	)
+	(
+		ini='typeset -ra -i arr=((10 11 12) 3 4)'
+		chg='(( arr[0][1] /= 2 ))'
+		res='typeset -p arr'
+		exp='typeset -r -a -i arr=((10 11 12) 3 4)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=10 [b]=20 [c]=30)'
+		chg='arr[b]=40'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=10 [b]=20 [c]=30)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=10 [b]=20 [c]=30)'
+		chg='arr[b]=()'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=10 [b]=20 [c]=30)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=10 [b]=20 [c]=30)'
+		chg='arr[b]=(40)'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=10 [b]=20 [c]=30)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+		chg='arr[c]=30'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+		chg='arr[c]+=30'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+		chg='arr[a][11]=0xb'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+		chg='arr[a][11]+=0xb'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+	)
+	(
+		ini='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+		chg='arr[a][13]=XIII'
+		res='typeset -p arr'
+		exp='typeset -r -A arr=([a]=([10]=X [11]=XI [12]=XII) [b]=20)'
+	)
+
+# Below tests cause a 'recursion too deep' error condition
```

Comment:
This makes no sense -- it's some kind of strange bug in 93u+ and 93u+m. There is no recursion here! I tried uncommenting those tests and running them on various ksh versions. The error does not occur when running the tests on 93v- or ksh2020.

Reply:
I figured it was a bug with arrays, compound variables, and/or arithmetic operators (#168) as I can manually run them. Hopefully, once that and other array issues get resolved those tests can be enabled.

Reply:
I did some debugging and looked at the differences between 93u+ and 93v- in streval.c which handles arithmetic expansion. The [DEBUG] output shows where the recursion level is spuriously increased, which is informative when running your tests. The [DEBUG] output on 93v- is identical, so nothing changed there, except that they simply changed `MAXLEVEL` from 9 to 1024 (!) in that version, which avoids the problem. Fixing this properly is for another commit, but for now I think you should increase `MAXLEVEL` to 1024 as in 93v- and enable the rest of the tests.
```diff
--- a/src/cmd/ksh93/sh/streval.c
+++ b/src/cmd/ksh93/sh/streval.c
@@ -47,7 +47,7 @@
 #   define SH_DICT	"libshell"
 #endif
 
-#define MAXLEVEL	9
+#define MAXLEVEL	1024
 #define SMALL_STACK	12
 
 /*
@@ -177,6 +177,7 @@ Sfdouble_t	arith_exec(Arith_t *ep)
 	node.nosub = 0;
 	node.ptr = 0;
 	node.eflag = 0;
+error(ERROR_warn(0),"[DEBUG] level==%d",level);
 	if(level++ >=MAXLEVEL)
 	{
 		arith_error(e_recursive,ep->expr,ep->emode);
```

Reply:
oooOOOOooo. nice. I go test it now.

Reply:
Your DEBUG output is GOLD! I added in a little echo for the row and its initial setup. It is the `-i'`s that directly cause the level to increase which start on rtests[25].

```diff
diff --git a/src/cmd/ksh93/sh/streval.c b/src/cmd/ksh93/sh/streval.c
index 9a4e552..d573a03 100644
--- a/src/cmd/ksh93/sh/streval.c
+++ b/src/cmd/ksh93/sh/streval.c
@@ -47,7 +47,7 @@
 #   define SH_DICT     "libshell"
 #endif

-#define MAXLEVEL       9
+#define MAXLEVEL       1024
 #define SMALL_STACK    12

 /*
@@ -177,6 +177,7 @@ Sfdouble_t  arith_exec(Arith_t *ep)
        node.nosub = 0;
        node.ptr = 0;
        node.eflag = 0;
+error(ERROR_warn(0),"[DEBUG] level==%d",level);
        if(level++ >=MAXLEVEL)
        {
                arith_error(e_recursive,ep->expr,ep->emode);
diff --git a/src/cmd/ksh93/tests/readonly.sh b/src/cmd/ksh93/tests/readonly.sh
index e7e9756..d1d43e7 100644
--- a/src/cmd/ksh93/tests/readonly.sh
+++ b/src/cmd/ksh93/tests/readonly.sh
@@ -266,12 +266,12 @@ rtests=(
        )

 # Below tests cause a 'recursion too deep' error condition
-#      (
-#              ini='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
-#              chg='(( arr[b] += 5 ))'
-#              res='typeset -p arr'
-#              exp='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
-#      )
+       (
+               ini='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
+               chg='(( arr[b] += 5 ))'
+               res='typeset -p arr'
+               exp='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
+       )
 #      (
 #              ini='typeset -r -A -i arr=([a]=10 [b]=20 [c]=30)'
 #              chg='(( arr[b] -= 5 ))'
@@ -321,6 +321,7 @@ typeset -i i
 n=${#rtests[@]}
 for ((i=0; i<$n; i++))
 do
+       echo $i: "${rtests[$i].ini}"
        got=$(
                trap "${rtests[$i].res}" EXIT
                eval "${rtests[$i].ini}"
```
```
24: typeset -ra arr=((a b c) 1)
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==0
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==0
25: typeset -ra -i arr=((10 11 12) 3 4)
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==1
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==1
26: typeset -ra -i arr=((10 11 12) 3 4)
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==2
/home/hyenias/ksh93/src/cmd/ksh93/tests/readonly.sh: line 322: warning: [DEBUG] level==2
```

Reply:
All done.